src/regdesk/regdesk.js

```diff
@@ -0,0 +1,106 @@
+</div>
+</div>
+<div class="d-flex flex-row justify-content-end">
+<div class="btn-group" role="group" aria-label="Printer button group">
+    <button id="printto_${index}" class="btn btn-success btn-lg" data-label-width="${printer.labelWidth}" data-label-height="${printer.labelHeight}" data-device-id="${index}">🖨</button>
+        <button class="btn btn-success dropdown-toggle dropdown-toggle-split" data-bs-toggle="dropdown" aria-expanded="false">
+            <span class="visually-hidden">Settings</span>
+        </button>
+        <ul class="dropdown-menu">
+            <li><a id="printtest_${index}" data-device-id="${index}" class="dropdown-item" href="#">
+                Print test page
+            </a></li>
+            <li><a id="autosense_${index}" data-device-id="${index}" class="dropdown-item" href="#">
+                AutoSense length
+            </a></li>
+            <li><a id="configprinter_${index}" data-device-id="${index}" class="dropdown-item" href="#">
+                Set label config
+            </a></li>
+            <li><a id="cutprinter_${index}" data-device-id="${index}" class="dropdown-item" href="#">
+                Cut label
+            </a></li>
+        </ul>
+    </div>
+</div>
+</div>
+</li>`;
+    document.getElementById("printerlist").appendChild(element);
+
+    document.getElementById(`printto_${index}`)
+        .addEventListener('click', async (e) => {
+            e.preventDefault();
+            let deviceId = e.target.dataset.deviceId;
+            printerMgr.printForm(deviceId);
+        });
+
+    document.getElementById(`autosense_${index}`)
+        .addEventListener('click', async (e) => {
+            e.preventDefault();
+            let deviceId = e.target.dataset.deviceId;
+            await printerMgr.printers[deviceId].configLabelWidth();
+            await printerMgr.printers[deviceId].setLabelHeightCalibration();
+        });
+
+    document.getElementById(`printtest_${index}`)
+        .addEventListener('click', async (e) => {
+            e.preventDefault();
+            let deviceId = e.target.dataset.deviceId;
+            await printerMgr.printers[deviceId].printTestPage();
+        });
+
+    document.getElementById(`cutprinter_${index}`)
+        .addEventListener('click', async (e) => {
+            e.preventDefault();
+            let deviceId = e.target.dataset.deviceId;
+            await printerMgr.printers[deviceId].addCmd("C").print();
+        });
+}
+
+////////////////////////////////
+// End most of the test code, though the rest of this will change too.
+
+
+var printerMgr;
+
+document.addEventListener('readystatechange', async () => {
+    if (document.readyState === "complete") {
+
+        const printerList = document.getElementById("printerlist");
+        printerMgr = new PrinterManager();
+
+        document.getElementById("addprinter").addEventListener('click', async () => {
+            await printerMgr.printerPair();
```

Comment:
This feels a little disjointed. Like there are two mechanisms to establishing communication with a printer. The navigator.usb.addEventListener (automatic? physical?) and this printerPair (manual? configuration?) and they behave differently.

Put another way, its almost like the "old" way of inserting a floppy. You would do the physical action of insertion, then go to your command prompt and type "mount  /mnt/floppy". Compared to today, where you do the physical action of insertion, then your pc pops up with an autoplay dialog. The latter is a lot nicer.

Reply:
So as you found eventually there are, indeed, two 'steps' of connecting to a printer. The first is granting the page access to talk to the printer _at all_, a necessarily manual process, and the actual connection thereof.

If the permission has been given in the past we can skip that step and jump right into the connect operation.